io/fasta/fasta.go

```diff
@@ -66,14 +70,178 @@ type Fasta struct {
 	}
-	return outputFastas, nil
+}
+
+// ParseAll parses all sequences in underlying reader only returning non-EOF errors.
+// It returns all valid fasta sequences up to error if encountered.
+func (p *Parser) ParseAll() ([]Fasta, error) {
+	return p.ParseN(math.MaxInt)
+}
+
+// ParseN parses up to maxSequences fasta sequences from the Parser's underlying reader.
+// ParseN does not return EOF if encountered.
+// If an non-EOF error is encountered it returns it and all correctly parsed sequences up to then.
+func (p *Parser) ParseN(maxSequences int) (fastas []Fasta, err error) {
+	for counter := 0; counter < maxSequences; counter++ {
+		fasta, _, err := p.ParseNext()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				err = nil // EOF not treated as parsing error.
+			}
+			return fastas, err
+		}
+		fastas = append(fastas, fasta)
+	}
+	return fastas, nil
+}
+
+// ParseByteLimited parses fastas until byte limit is reached.
+// This is NOT a hard limit. To set a hard limit on bytes read use a
+// io.LimitReader to wrap the reader passed to the Parser.
+func (p *Parser) ParseByteLimited(byteLimit int64) (fastas []Fasta, bytesRead int64, err error) {
+	for bytesRead < byteLimit {
+		fasta, n, err := p.ParseNext()
+		bytesRead += n
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				err = nil // EOF not treated as parsing error.
+			}
+			return fastas, bytesRead, err
+		}
+		fastas = append(fastas, fasta)
+	}
+	return fastas, bytesRead, nil
+}
+
+// ParseNext reads next fasta genome in underlying reader and returns the result
+// and the amount of bytes read during the call.
+// ParseNext only returns an error if it:
+//   - Attempts to read and fails to find a valid fasta sequence.
+//   - Returns reader's EOF if called after reader has been exhausted.
+//   - If a EOF is encountered immediately after a sequence with no newline ending.
+//     In this case the Fasta up to that point is returned with an EOF error.
+//
+// It is worth noting the amount of bytes read are always right up to before
+// the next fasta starts which means this function can effectively be used
+// to index where fastas start in a file or string.
+func (p *Parser) ParseNext() (Fasta, int64, error) {
+	if _, err := p.rd.Peek(1); err != nil {
+		// Early return on error. Probably will be EOF.
+		return Fasta{}, 0, err
+	}
+	// Initialization of parser state variables.
+	var (
+		// Parser looks for a line starting with '>' (U+003E)
+		// that contains the next fasta sequence name.
+		lookingForName = true
+		seqName        string
+		sequence, line []byte
+		err            error
+		totalRead      int64
+	)
+	const translationStopChar = '*'
+	for {
+		line, err = p.rd.ReadSlice('\n')
+		totalRead += int64(len(line))
+		p.line++
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				if len(line) > 1 && line[0] != ';' {
+					// EOF-ended fasta. We append line if not empty
+					stopTranslate := bytes.IndexByte(line, translationStopChar)
```

Comment:
The stop codon only denotes that a gene will stop being transcribed. In practice a single sequence can have many stop codons. Is it safe to remove this block @soypat?

Reply:
Oh snap, right will do. Just pushed a incorrect solution in that case